pvlib/modelchain.py

```diff
@@ -632,31 +716,44 @@ def infer_dc_model(self):
                              'set the model with the dc_model kwarg.')
 
     def sapm(self):
-        self.dc = self.system.sapm(self.effective_irradiance,
-                                   self.cell_temperature)
+        self.results.dc = self.system.sapm(self.results.effective_irradiance,
+                                           self.results.cell_temperature)
 
-        self.dc = self.system.scale_voltage_current_power(self.dc)
+        self.results.dc = self.system.scale_voltage_current_power(
+            self.results.dc)
 
         return self
 
     def _singlediode(self, calcparams_model_function):
-        (photocurrent, saturation_current, resistance_series,
-         resistance_shunt, nNsVth) = (
-            calcparams_model_function(self.effective_irradiance,
-                                      self.cell_temperature))
-
-        self.diode_params = pd.DataFrame({'I_L': photocurrent,
-                                          'I_o': saturation_current,
-                                          'R_s': resistance_series,
-                                          'R_sh': resistance_shunt,
-                                          'nNsVth': nNsVth})
-
-        self.dc = self.system.singlediode(
-            photocurrent, saturation_current, resistance_series,
-            resistance_shunt, nNsVth)
-
-        self.dc = self.system.scale_voltage_current_power(self.dc).fillna(0)
-
+        def _make_diode_params(photocurrent, saturation_current,
+                               resistance_series, resistance_shunt,
+                               nNsVth):
+            return pd.DataFrame(
+                {'I_L': photocurrent, 'I_o': saturation_current,
+                 'R_s': resistance_series, 'R_sh': resistance_shunt,
+                 'nNsVth': nNsVth}
+            )
+        params = calcparams_model_function(self.results.effective_irradiance,
+                                           self.results.cell_temperature)
+        if self.system.num_arrays == 1:
```

Comment:
we have to handle singletons vs iterables here because of the `_unwrap_single_value` decorator, right? Kind of awkward here, so perhaps would be awkward in user code too.

Reply:
That's probably 2/3 true most of the complexity comes from `_unwrap_single_value`, but part of it comes from trying to mirror that same behavior in `ModelChain`. I snuck an optional kwarg into the decorator that can prevent unwrapping (`unwrap=False`), but we would then need to unwrap inside `ModelChain` to maintain the same behavior when users want `ModelChain.dc` for a single array system.

The other problem with the "unwrap" parameter (and with changing the type passed around internally) is that because the ModelChain tests are built with mocks we would need to change at least some of the existing tests to handle the difference in how the `PVSystem` functions are called.

It does have the potential to be awkward in user code, but I think the typical use case is that you are modeling a specific system so the type will be consistent within your code. I'm not sure of a way around that as long as we want to maintain the "legacy" single-array behavior.

Reply:
Here is what it could look like using `unwrap=False`

```python
    def _singlediode(self, calcparams_model_function):
        def _make_diode_params(photocurrent, saturation_current,
                               resistance_series, resistance_shunt,
                               nNsVth):
            return pd.DataFrame(
                {'I_L': photocurrent, 'I_o': saturation_current,
                 'R_s': resistance_series, 'R_sh': resistance_shunt,
                 'nNsVth': nNsVth}
            )
        params = calcparams_model_function(self.results.effective_irradiance,
                                           self.results.cell_temperature,
                                           unwrap=False)
        self.results.diode_params = tuple(itertools.starmap(
            _make_diode_params, params))
        self.results.dc = tuple(itertools.starmap(
            self.system.singlediode, params))
        self.results.dc = self.system.scale_voltage_current_power(
            self.results.dc,
            unwrap=False
        )
        self.results.dc = tuple(dc.fillna(0) for dc in self.results.dc)
        if self.system.num_arrays == 1:
            self.results.diode_params = self.results.diode_params[0]
            self.results.dc = self.results.dc[0]
        return self
```

Reply:
> It does have the potential to be awkward in user code, but I think the typical use case is that you are modeling a specific system so the type will be consistent within your code. 

Maybe, but I can see myself modeling some systems with a single Array and some systems with multiple Arrays. But I'd also be using ModelChain in that situation, so I will stop complaining!

> I'm not sure of a way around that as long as we want to maintain the "legacy" single-array behavior.

I agree. It might be possible with broadcasting if we were only considering numpy arrays but pandas makes this a lot harder. I think we should continue with this approach for now.

> Here is what it could look like using unwrap=False

That's more clear to me but I suppose others may have a different view.

Reply:
I also think that is more clear. I'd like to use it, but I'm getting tripped up by a test failure from `mocker.spy()` for `test_infer_dc_model`. It is complaining about the `unwrap` kwarg since it isn't part of the "true" signature. What is odd is that the test only fails for `pvsyst`, the other models that use `ModelChain._singlediode()` internally (`cec` and `desoto`) pass the test.

Reply:
I dug in to this further and discovered why only `pvsyst` fails: `PVSystem.calcparams_desoto` and `PVSystem.calcparams_cec`  both have a `**kwargs` param; however, it looks like it is not used (obvious in this branch, but more subtle in `master`). Here is `PVSystem.calcparams_cec` as it exists in master:

```python
    def calcparams_cec(self, effective_irradiance, temp_cell, **kwargs):
        """
        Use the :py:func:`calcparams_cec` function, the input
        parameters and ``self.module_parameters`` to calculate the
        module currents and resistances.
        Parameters
        ----------
        effective_irradiance : numeric
            The irradiance (W/m2) that is converted to photocurrent.
        temp_cell : float or Series
            The average cell temperature of cells within a module in C.
        **kwargs
            See pvsystem.calcparams_cec for details
        Returns
        -------
        See pvsystem.calcparams_cec for details
        """

        kwargs = _build_kwargs(['a_ref', 'I_L_ref', 'I_o_ref', 'R_sh_ref',
                                'R_s', 'alpha_sc', 'Adjust', 'EgRef', 'dEgdT',
                                'irrad_ref', 'temp_ref'],
                               self.module_parameters)

        return calcparams_cec(effective_irradiance, temp_cell, **kwargs)
```

`kwargs` gets shadowed. If the keyword args provided to the method are meant to be shadowed we should just drop the `**kwargs` parameter, if they are supposed to be used then this is probably a bug. Because they can take arbitrary kwargs `mocker.spy()` doesn't complain about the `unwrap=True` parameter.

Given that it would require re-thinking the core testing strategy for `ModelChain` we may want to stick with the slightly harder to read, but mockable implementation of this method.

Reply:
We should remove `**kwargs` from those methods. Options for that: 

1. a separate PR to deprecate them in 0.8.1 and mark for removal in 0.9
2. after 0.8.1, a separate PR to remove them in 0.9 without deprecation
2. after 0.8.1, remove them in this PR without deprecation.

I looked into modifying the mock spec or the wrap spec but couldn't figure out a good way to do either. The most straightforward thing might be to move the mock down to the function layer. Might just be a matter of changing  `mocker.spy(system, dc_model_function[dc_model])` to `mocker.spy(pvsystem, dc_model_function[dc_model])`. Probably would need to do that for the other model steps. I don't know if that's a net benefit or not.

Reply:
It did not occur to me to move the mock to the function layer, I've been all-consumed by the object layer. That would probably work. We would need to do that for any tests that mock `PVSystem` methods and cover code that uses `unwrap=False`. For `ModelChain._singlediode()` there is only one such test.